### PR TITLE
Towards API/ABI stability + refinement of AcceleratorId & ContextFlags

### PR DESCRIPTION
**Pull request by m4rs-mt**
We are aiming for API/ABI stability in v1.0. Therefore it is a good time to replace the (mostly obsolete) enumeration `ContextFlags`. In addition, there were also many questions and issue reports from the community about supported and detected accelerators (especially related to the `OpenCL` accelerator, like #424). Related to these detection issues, there are also some reports about freezing calls to the `OpenCL` runtime API even though users only want to use the `CPU`/`Cuda` accelerators.

This PR removes the enumeration `ContextFlags` and implements the same functionality in an object oriented way using a `Context.Builder` class. It offers a fluent-API like configuration interface which makes it easy to set up:
```c#
// Enables all supported accelerators (default CPU accelerator only) and puts the context
// into auto-assertion mode via "AutoAssertions()". In other words, if a debugger is attached,
// the `Context` instance will turn on all assertion checks. This behavior is identical
// to the current implementation via new Context();
using var context = Context.CreateDefault();

// Turns on O2 and enables all compatible Cuda devices.
using var context = Context.Create(builder =>
{
    builder.Optimize(OptimizationLevel.O2).Cuda();
});

// Turns on all assertions, enables the IR verifier and enables all compatible OpenCL devices.
using var context = Context.Create(builder =>
{
    builder.Assertions().Verify().OpenCL();
});

// Turns on kernel source-line annotations, fast math using 32-bit float and enables
// *all* (even incompatible) OpenCL devices.
using var context = Context.Create(builder =>
{
    builder
        .DebugSymbols(DebugSymbolsMode.KernelSourceAnnotations)
        .Math(MathMode.Fast32BitOnly)
        .OpenCL(device => true);
});

// Selects an OpenCL device with a warp size of at least 32:
using var context = Context.Create(builder =>
{
    builder.OpenCL(device => device.WarpSize >= 32);
});

// Turns on all assertions in debug mode (same behavior like calling CreateDefault()):
using var context = Context.Create(builder =>
{
    builder.AutoAssertions();
});

// Turns on debug optimizations (level O0) and all assertions if a debugger is attached:
using var context = Context.Create(builder =>
{
    builder.AutoDebug();
});

// Turns on debug mode (optimization level P0, assertions and kernel debug information):
using var context = Context.Create(builder =>
{
    builder.Debug();
});

// Disable caching, enable conservative inlining and inline mutable static field values:
using var context = Context.Create(builder =>
{
    builder
        .Caching(CachingMode.Disabled)
        .Inlining(InliningMode.Conservative)
        .StaticFields(StaticFieldMode.MutableStaticFields);
});

// Turn on *all* CPU accelerators that simulate different hardware platforms:
using var context = Context.Create(builder => builder.CPU());

// Turn on an AMD-based CPU accelerator:
using var context = Context.Create(builder => builder.CPU(CPUDeviceKind.AMD));
```
Note that by default all debug symbols are automatically turned off when a debugger is attached. If you want to turn on the debug information in all cases, call `.builder.DebugSymbols(DebugSymbolsMode.Basic)`. At the same time, this PR introduces the notion of a `Device`, which replaces the implementation of `AcceleratorId`. This allows us to query detailed device information without explicitly instantiating an accelerator:
```c#
// Print all device information without instantiating a single accelerator
// (device context) instance.
using var context = Context.Create(...);
foreach (var device in context)
{
    // Print detailed accelerator information
    device.PrintInformation();

    // ...
}
```

Note that this PR removes the ability to call the accelerator constructors (e.g. `new CudaAccelerator(...)`) directly. Either use the `CreateAccelerator` methods defined in the `Device` classes or use one of the extension methods like `CreateCudaAccelerator(...)` of the `Context` class itself:
```c#
using var context = Context.Create(...);
foreach (var device in context)
{
    // Instantiate an accelerator instance on this device
    using Accelerator accel = device.CreateAccelerator();
    // ...
}

// Instantiate the 2nd Cuda accelerator (NOTE that this is the *2nd* Cuda device
// and *not* the 2nd device of your machine).
using CudaAccelerator cudaDevice = context.CreateCudaAccelerator(1);

// Instantiate the 1st OpenCL accelerator (NOTE that this is the *1st* OpenCL device
// and *not* the 1st device of your machine).
using CLAccelerator clDevice = context.CreateOpenCLAccelerator(0);
```

`Context` properties that expose types from other (ILGPU internal) namespaces that cannot/should not (?) be covered by the API/ABI guarantees we want to give, has been made `internal` properties. To access these properties, use one of the available extension methods located in the corresponding namespaces:
```c#
using var context = ...

// OLD way
var internalIRContext = context.IRContext;

// NEW way:
// using namespace ILGPU.IR;
var internalIRContext = context.GetIRContext();
```

Note that PR #427 should be merged first.

### CODE REVIEW

**Review by MoFtZ — CHANGES_REQUESTED** [review]

File: Src/ILGPU/Runtime/Cuda/CudaDevice.cs [hunk 1]
```diff
@@ -0,0 +1,548 @@
...
+        /// <summary>
+        /// Constructs a new Cuda accelerator reference.
+        /// </summary>
+        /// <param name="deviceId">The Cuda device id.</param>
+        internal CudaDevice(int deviceId)
+        {
+            if (deviceId < 0)
+                throw new ArgumentOutOfRangeException(nameof(deviceId));
+
+            Backend.EnsureRunningOnPlatform(TargetPlatform.X64);
```
[the remark this answers is not in the document]
Comment (m4rs-mt) [reply Src/ILGPU/Runtime/Cuda/CudaDevice.cs:113]: @MoFtZ You are right. We have already introduced the UVA features for unified addresses, which work only on 64-bit machines. Therefore, we should disable the X86 Cuda support.

File: Src/ILGPU/Runtime/Accelerator.cs [hunk 2]
```diff
@@ -145,11 +123,14 @@ public static AcceleratorType CurrentType
         /// Constructs a new accelerator.
         /// </summary>
         /// <param name="context">The target context.</param>
-        /// <param name="type">The target accelerator type.</param>
-        internal Accelerator(Context context, AcceleratorType type)
+        /// <param name="device">The device.</param>
+        protected Accelerator(Context context, Device device)
```
[the remark this answers is not in the document]
Comment (m4rs-mt) [reply Src/ILGPU/Runtime/Accelerator.cs:127]: I am not sure. An `Accelerator` implements `IDevice` in order to share common properties and extension methods with the `Device` class. However, an accelerator should not be instantiated using another `Accelerator` to my mind.

File: Src/ILGPU/ContextProperties.cs [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+    /// </summary>
+    public enum InliningMode
+    {
+        /// <summary>
+        /// All functions will be inlined by default.
+        /// </summary>
+        /// <remarks>
+        /// This is the default setting.
+        /// </remarks>
+        Default,
```
[the remark this answers is not in the document]
Comment (m4rs-mt) [reply Src/ILGPU/ContextProperties.cs:17]: The problem with these enumeration values in general is that we cannot change them again due to ABI compatibility constraints. Maybe we should rename `Default` to `Aggressive`.

File: Src/ILGPU/Backends/PTX/PTXArchitecture.cs [hunk 4]
```diff
@@ -284,20 +284,35 @@ public static class PTXArchitectureUtils
...
         #region Static
 
+        /// <summary>
+        /// Tries to get the PTX architecture for the given major and minor versions.
+        /// </summary>
+        /// <param name="major">The major version.</param>
+        /// <param name="minor">The minor version.</param>
+        /// <param name="architecture">The determined architecture (if any).</param>
+        /// <returns>True, if the architecture is supported.</returns>
+        public static bool TryGetArchitecture(
```
[the remark this answers is not in the document]
Comment (m4rs-mt) [reply Src/ILGPU/Backends/PTX/PTXArchitecture.cs:294]: Makes sense. I am going to move them to the `Cuda` namespace.

File: Src/ILGPU/Backends/PTX/PTXArchitecture.cs [hunk 5]
```diff
@@ -284,20 +284,35 @@ public static class PTXArchitectureUtils
...
+            int major,
+            int minor,
+            out PTXArchitecture architecture)
+        {
+            architecture = new PTXArchitecture(major, minor);
+            return
+                architecture >= PTXArchitecture.SM_30 &&
+                architecture <= PTXArchitecture.SM_86;
```
[the remark this answers is not in the document]
Comment (m4rs-mt) [reply Src/ILGPU/Backends/PTX/PTXArchitecture.cs:302]: What do you suggest? Skip the second check against `SM_86`?

File: Src/ILGPU/Backends/PTX/PTXBackend.cs [hunk 6]
```diff
@@ -144,13 +161,13 @@ public sealed class PTXBackend :
             // Ensure that all intrinsics can be generated
             backendContext.EnsureIntrinsicImplementations(IntrinsicProvider);
 
-            bool useDebugInfo = Context.HasFlags(
-                ContextFlags.EnableKernelDebugInformation);
+            var debugSymbolsMode = Context.Properties.DebugSymbolsMode;
+            bool useDebugInfo = debugSymbolsMode > DebugSymbolsMode.Kernel;
```
[the remark this answers is not in the document]
Comment (m4rs-mt) [reply Src/ILGPU/Backends/PTX/PTXBackend.cs:165]: Nice one 👍

File: Src/ILGPU/Context.Builder.cs [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+            }
+
+            #endregion
+
+            #region Properties
+
+            /// <summary>
+            /// All accelerator descriptions that have been registered.
+            /// </summary>
+            public DeviceRegistry AcceleratorRegistry { get; } =
```
[the remark this answers is not in the document]
Comment (m4rs-mt) [reply Src/ILGPU/Context.Builder.cs:57]: 👍 let's change its visibility to `internal`

File: Src/ILGPU/Backends/PTX/PTXBackend.cs [hunk 8]
```diff
@@ -144,13 +161,13 @@ public sealed class PTXBackend :
...
             PTXDebugInfoGenerator debugInfoGenerator = PTXNoDebugInfoGenerator.Empty;
             if (useDebugInfo)
             {
                 debugInfoGenerator =
-                    Context.HasFlags(ContextFlags.EnableInlineSourceAnnotations)
+                    debugSymbolsMode > DebugSymbolsMode.KernelSourceAnnotations
```
[the remark this answers is not in the document]
Comment (m4rs-mt) [reply Src/ILGPU/Backends/PTX/PTXBackend.cs:170]: 👍

File: Src/ILGPU/Context.Builder.cs [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+            /// <summary>
+            /// Specifies the debug symbol mode to use.
+            /// </summary>
+            /// <param name="debugSymbolsMode"></param>
+            /// <returns></returns>
+            public Builder DebugSymbols(DebugSymbolsMode debugSymbolsMode)
+            {
+                DebugSymbolsMode = (DebugSymbolsMode)System.Math.Max(
+                    (int)DebugSymbolsMode,
+                    (int)debugSymbolsMode);
```
[the remark this answers is not in the document]
Comment (m4rs-mt) [reply Src/ILGPU/Context.Builder.cs:119]: QUESTION: What should happen if someone invokes `Assertions()` and then disables debug information via `DebugSymbols(Disabled)`?

File: Src/ILGPU/Backends/PTX/PTXArchitecture.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (MoFtZ) [reply Src/ILGPU/Backends/PTX/PTXArchitecture.cs]: Yes, I'm not sure the second check is necessary. As long as the architecture meets the minimum requirement, are there any restrictions on using future/unknown architectures?

File: Src/ILGPU/Backends/PTX/PTXArchitecture.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (MoFtZ) [reply Src/ILGPU/Backends/PTX/PTXArchitecture.cs]: Should this also apply to `PTXInstructionSet`?

File: Src/ILGPU/Context.Builder.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (MoFtZ) [reply Src/ILGPU/Context.Builder.cs:118]: Yes, it is an interesting problem. Some possible options are:
1. Last action wins, order of builder calls is important.
1. Runtime exception, informing the caller of a conflict.
1. Highest mode wins, as per this PR.

I'm not sure there is a "correct" answer here.

File: Src/ILGPU/Backends/PTX/PTXArchitecture.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (m4rs-mt) [reply Src/ILGPU/Backends/PTX/PTXArchitecture.cs]: Makes sense to me. I have removed the second check.

File: Src/ILGPU/Backends/PTX/PTXArchitecture.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (m4rs-mt) [reply Src/ILGPU/Backends/PTX/PTXArchitecture.cs]: I think you are right: we should also rename `PTXInstructionSet` to `CudaInstructionSet` and move the `Runtime` namespace. This allows us to "keep private" the backend namespace.

File: Src/ILGPU/Context.Builder.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (m4rs-mt) [reply Src/ILGPU/Context.Builder.cs:118]: Perhaps we should throw an `InvalidOperationException` indicating a conflicting situation.

**Pull request merged**